### ISSUE TITLE
What was the "element is being created" flag for?

### ISSUE THREAD

**Issue by domenic** [WICG/webcomponents#458]
@dglazkov, need help again.

Previous revisions of the spec, when createdCallback existed, had this text under createdCallback:

> For the duration of this callback invocation, the element is being created flag MUST be set to true. In all other cases, the flag MUST be set to false.
> 
> If the created callback exists for an element, all other callbacks MUST NOT be enqueued until after this created callback's invocation had started.

Furthermore, both previous revisions and the current spec have a clause:

> If element's element is being created flag is false, add element to the current element queue.

inside the enqueuing of custom element callbacks. That is, if the element is being created:
- The relevant reaction _will_ be added to element's custom element reaction queue, but
- element _will not_ be added to the current element queue (i.e. to the queue on the top of the stack-of-queues)

I am trying to figure out what the intent was here, and whether we need to reintroduce setting "element is being created" before/after calling the constructor in the current design, or whether we should remove the "element is being created" flag and the resulting fork in the enqueuing logic.

**Comment by domenic** [comment]
It's also worth noting that I'm not sure we can make this flag work at all for `new XTag()`, so whatever it was trying to accomplish might not be possible...

**Comment by dglazkov** [comment]
This was to prevent situations where other callbacks could sneak in before createdCallback had completed: https://www.w3.org/Bugs/Public/show_bug.cgi?id=22966

**Comment by domenic** [comment]
Perfect background reading; thank you.

So the goal here, in modern terms, is that

``` js
class XTag extends HTMLElement {
  constructor() {
    super();
    console.log("constructor start");
    this.setAttribute("foo", "bar");
    console.log("constructor end");
  }

  static get observedAttributes() { return ["foo"]; }

  attributeChangedCallback() {
    console.log("attribute changed");
  }
}

customElements.define("x-tag", XTag);

new XTag();
```

should log:

```
constructor start
constructor end
attribute changed
```

As I said above, I can't really think of any way to make this work in in the specific case of `new XTag()`, since (unlike in the previous design) we don't control the constructor code, and thus don't know when to set/unset an "element is being created" flag. In other words, there's no way for `setAttribute` to know that it should delay calling `attributeChangedCallback`, and even if we could let it know, it wouldn't know _when_ to delay too.

That's kind of sad.

We could make this work for `createElement` and the parser and friends, but I assume that the inconsistency there would not be a good idea.

On the upside, these days we're very explicitly telling people that they shouldn't be mutating the DOM in their constructor. So, if developers follow the conformance requirements, then this shouldn't come up very often, which is good...

In conclusion I think we'll have to "regress" on this behavior (so the above outputs `constructor start / attribute changed / constructor end`). I'll close this with a commit that removes the "element is being created" flag. If this is a big deal to anyone, please let me know and we can reopen for further discussion (although I'm not really sure what we could do). Tagging in @rniwa @annevk @travisleithead as browser reps, and @coonsta and @sjmiles since they were in the original thread at https://www.w3.org/Bugs/Public/show_bug.cgi?id=22966.

**Comment by annevk** [comment]
Can't you do this by setting this flag before you invoke the constructor. Then once the constructor has returned, you unset it and you check if you have any queued callbacks. Which you then iterate over and call?

**Comment by domenic** [comment]
Who sets the flag? We don't know when the constructor is invoked. Author code could do so at any time.

**Comment by annevk** [comment]
Ah, I was just thinking about upgrade scenarios. We could make it work for those. But yes, straight new XElement would fail. I guess it's fine.

**Comment by domenic** [comment]
Yeah, we could make it work for all the scenarios where we're calling the constructor---upgrades, createElement, and the parser. But having the ordering be different in those scenarios vs. `new XTag()` seems pretty bad.
